docs: update tutorials and how-tos to use pytest-jubilant (#2406)
canonical/operator: docs/howto/migrate/migrate-integration-tests-from-pytest-operator.md

(pytest-operator-migration)=
# How to migrate integration tests from pytest-operator

Older charm integration tests use [pytest-operator](https://github.com/charmed-kubernetes/pytest-operator) and [python-libjuju](https://github.com/juju/python-libjuju). This guide explains how to migrate your integration tests from those libraries to Jubilant and [`pytest-jubilant`](https://github.com/canonical/pytest-jubilant).

```{tip}
Try bootstrapping your migration with an AI Agent (such as GitHub Copilot or Claude Code). Instruct the agent to clone the `canonical/jubilant` and `canonical/pytest-jubilant` repositories, study them, and then migrate the charm integration tests to Jubilant. You should end up with a great starting point to then continue as outlined in the rest of this guide.
```

To get help while you're migrating tests, please keep the {external+jubilant:doc}`Jubilant API Reference <reference/jubilant>` handy, and make use of your IDE's autocompletion -- Jubilant tries to provide good type annotations and docstrings.

Migrating your tests can be broken into three steps:

1. Update your dependencies

> 2. Provide the resources your tests need

Reviewer: Wouldn't they still need to add a fixture for getting the packed charm? Although I suppose not if they're using a Charmcraft profile. So I'm good with your change.

Author: Yes, step 2 is now "write one or more charm fixtures" + "use the pytest-jubilant fixtures".

(pytest-operator-migration)=
# How to migrate integration tests from pytest-operator

> Many charm integration tests use [pytest-operator](https://github.com/charmed-kubernetes/pytest-operator) and [python-libjuju](https://github.com/juju/python-libjuju). This guide explains how to migrate your integration tests from those libraries to [`pytest-jubilant`](https://github.com/canonical/pytest-jubilant), which uses the Jubilant library.

Reviewer 2: Maybe it's a good time to update the "many"? Also, I don't think we should say that you're migrating to pytest-jubilant; it always bugged me that people would say that they were using pytest-operator for their integration tests, when it reality they were using python-libjuju and a bit of wrapping from pytest-operator.

```suggestion
Older charm integration tests use [pytest-operator](https://github.com/charmed-kubernetes/pytest-operator) and [python-libjuju](https://github.com/juju/python-libjuju). This guide explains how to migrate your integration tests from those libraries to Jubilant and [`pytest-jubilant`](https://github.com/canonical/pytest-jubilant).
```

     boto3
     cosl
-    juju>=3.0
+    jubilant>=1.8,<2
+    pytest-jubilant>=2,<3
     pytest
-    pytest-operator
-    pytest-asyncio
     -r{toxinidir}/requirements.txt
```

If you're migrating a large number of tests, you may want to do it in stages. In that case, keep the old dependencies in place till the end, and migrate tests one at a time, so that both `pytest-operator` and Jubilant tests can run together. Note that `pytest-operator` and `pytest-jubilant` use completely different CLI options, so make sure you provide the correct ones for each if you're trying to do something fancy.

## Provide the resources your tests need

Your integration tests may use a combination of `pytest-operator` features for the resources they need, including packed charms, Juju models, and deployed applications. This section covers how to provide these resources when writing `Jubilant` based integration tests.

### Provide packed charms to your Python tests

`pytest-operator` provided a `build_charm` helper function. `pytest-jubilant` does not provide an equivalent helper, because it's cleaner to keep packing out of your Python integration tests.

> In CI, you may already follow a strategy of first packing your charms (in parallel), and then providing the packed charms to your (perhaps also parallelised) integration tests. A good way to provide the charms is via environment variables.

Reviewer: "via" 🙂 How about "with" instead?

Author: ```suggestion
In CI, you may already follow a strategy of first packing your charms (in parallel), and then providing the packed charms to your (perhaps also parallelised) integration tests. A good way to provide the charms is by using environment variables.
```

```ini
[testenv:integration]
pass_env =
    CHARM_PATH
commands =
    pytest --tb=native -vv --log-cli-level=DEBUG {toxinidir}/tests/integration {posargs}
```

In your integration tests themselves, you should define a fixture for your charm, which reads the environment variable if set, or falls back to looking for the packed charm in its expected location.

```py
# tests/integration/conftest.py
import os
import pathlib

@pytest.fixture(scope="session")
def charm_path():
    # Assuming the current working directory is the charm root:
    yield get_charm_path(env_var="CHARM_PATH", default_dir=pathlib.Path())

def get_charm_path(env_var: str, default_dir: pathlib.Path) -> pathlib.Path:
    charm = os.environ.get(env_var)
    if not charm:
        charms = list(default_dir.glob('*.charm'))
        assert charms, f'No charms were found in {default_dir}'
        assert len(charms) == 1, f'Found more than one charm {charms}'
        charm = charms[0]
    path = pathlib.Path(charm).resolve()
    assert path.is_file(), f'{path} is not a file'
    return path

> ```

Reviewer: Do you think we should use the same fixture in the how-to guide, in the "Write fixtures" section? How about in https://github.com/canonical/charmcraft/pull/2623?

Author: > How about in [canonical/charmcraft#2623](https://github.com/canonical/charmcraft/pull/2623)?

Per discussion, we've updated the templates to use a similar version (removing the layer of indirection of the helper function).

> Do you think we should use the same fixture in the how-to guide, in the "Write fixtures" section?

We probably should at least use the one from the template. What do you think, across the three locations (template, how-to, migration), which, if any, should use the multiple-charm-friendly helper function? Perhaps just the migration guide?

Author: Done

#### Packing multiple charms

> In a more complicated case where you have multiple charms, you should use `*CHARM_PATH` in `pass_env` instead, and use named environment variables to pass each charm's location (e.g. `FOO_CHARM_PATH`, `BAR_CHARM_PATH`).

Author: What I liked about `*CHARM_PATH` is that it also matches `CHARM_PATH`, which `*_CHARM_PATH` wouldn't. But I'm not sure which we should really prefer.

If you're migrating a large number of tests, you may want to do it in stages. In that case, keep the old dependencies in place till the end, and migrate tests one at a time, so that both pytest-operator and Jubilant tests can run together.

## Add any extra fixtures to `conftest.py`

> The `pytest-jubilant` plugin provides a module-scoped `juju` fixture that creates a temporary model, destroys it after the tests, and dumps debug logs on failure. It also provides CLI options such as `--no-juju-teardown` (to keep models) and `--juju-model` (to set a custom model name prefix). This replaces the need to write a `juju` fixture yourself.

Reviewer 2: I think if they are migrating from pytest-operator, they aren't expecting to write a fixture themselves, it would more be that it provides what OpsTest did.

```suggestion
The `pytest-jubilant` plugin provides a module-scoped `juju` fixture that creates a temporary model, destroys it after the tests, and dumps debug logs on failure. It also provides CLI options such as `--no-juju-teardown` (to keep models) and `--juju-model` (to set a custom model name prefix).
```

A few things to note about the fixture:

* To keep models around after running the tests (matching pytest-operator's `--keep-models`), pass `--no-juju-teardown`.
* To match python-libjuju's 10-minute `wait_for_idle` timeout, set `juju.wait_timeout = 10 * 60` in a wrapper fixture or at the start of your test.
* If any of the tests fail, the plugin automatically dumps the last 1000 lines of `juju debug-log` output.
* It is module-scoped, like pytest-operator's `ops_test` fixture. This means that a new model is created for every `test_*.py` file, but not for every test.

If your `test_*.py` module needs multiple Juju models (previously managed with `ops_test.track_model`), use the `juju_factory` fixture. This fixture lets you add additional models with their own unique suffixes -- no suffix is equivalent to the `juju` fixture.

```py
import jubilant
import pytest
import pytest_jubilant

@pytest.mark.fixture(scope="module")
def other_model(juju_factory: pytest_jubilant.JujuFactory):
    yield juju_factory.get_juju("other")

def test_cross_model(juju: jubilant.Juju, other_model: jubilant.Juju):
    ...
```

(how_to_migrate_an_application_fixture)=
### Application setup

> A lot of the time, you won't want to deploy your application in each test. In this case, you should use tests marked with `juju_setup`. These can be skipped in subsequent test runs using `--no-juju-setup` if you previously kept your models up and the applications deployed with `--no-juju-teardown`. This corresponds to `pytest-operator`'s `skip_if_deployed` functionality.

Author: Thanks for the suggestions, I agree that this is a bit confusing as-is. I'm rewriting as follows, let me know what you think:
```suggestion
A lot of the time, you won't want to deploy your application in each test. In this case, you should test deployment in the first tests in a module, and assume deployment was successful in subsequent tests.

It's a good idea to mark your deploy tests with `juju_setup`. If you use `--no-juju-teardown` to keep yoru models up and the applications deployed, then subsequent test runs can skip your `juju_setup` tests using `--no-juju-setup`. This corresponds to `pytest-operator`'s `skip_if_deployed` functionality.
```

* If any of the tests fail, the plugin automatically dumps the last 1000 lines of `juju debug-log` output.
* It is module-scoped, like pytest-operator's `ops_test` fixture. This means that a new model is created for every `test_*.py` file, but not for every test.

If your `test_*.py` module needs multiple Juju models (previously managed with `ops_test.track_model`), use the `juju_factory` fixture. This fixture lets you add additional models with their own unique suffixes -- no suffix is equivalent to the `juju` fixture.

```py
import jubilant
import pytest
import pytest_jubilant

@pytest.mark.fixture(scope="module")
def other_model(juju_factory: pytest_jubilant.JujuFactory):
    yield juju_factory.get_juju("other")

def test_cross_model(juju: jubilant.Juju, other_model: jubilant.Juju):
    ...
```

(how_to_migrate_an_application_fixture)=
### Application setup

A lot of the time, you won't want to deploy your application in each test. In this case, you should test deployment in the first tests in a module, and assume deployment was successful in subsequent tests.

> It's a good idea to mark your deploy tests with `juju_setup`. If you use `--no-juju-teardown` to keep yoru models up and the applications deployed, then subsequent test runs can skip your `juju_setup` tests using `--no-juju-setup`. This corresponds to `pytest-operator`'s `skip_if_deployed` functionality.

Author: ```suggestion
It's a good idea to mark your deploy tests with `juju_setup`. If you use `--no-juju-teardown` to keep your models up and the applications deployed, then subsequent test runs can skip your `juju_setup` tests using `--no-juju-setup`. This corresponds to `pytest-operator`'s `skip_if_deployed` functionality.
```
